README.md
#JSON API
A JSON API for WordPress

##Description
JSON API allows you to retrieve and manipulate WordPress content using HTTP requests. There are three main goals:

1. Provide a simple, consistent external interface
2. Create a stable, understandable internal implementation
3. Enable new types of extensions for WordPress

This plugin was created at [The Museum of Modern Art](http://moma.org/) for the weblog [Inside/Out](http://moma.org/explore/inside_out), which is served from Ruby on Rails. Instead of reimplementing the site templates as a WordPress theme, we opted for a Rails front-end that displays content served from a WordPress back-end. JSON API provides the necessary interface for retrieving content and accepting comment submissions.

##Installation
1. Upload the `json-api` folder to the `/wp-content/plugins/` directory or install directly through the plugin installer.
2. Activate the plugin through the 'Plugins' menu in WordPress or by using the link provided by the plugin installer.

##Table of Contents
 - [General concepts](#general-concepts)
   - [Requests](#requests)
   - [Controllers](#controllers)
   - [Responses](#responses)
 - [Request methods](#request-methods)
   - [Core controller methods](#core-controller-methods)
   - [Posts controller methods](#pages-controller-methods)
   - [Respond controller methods](#respond-controller-methods)
   - [Widgets controller methods](#widgets-controller-methods)
 - [Request arguments](#request-arguments)
   - [Output-modifying arguments](#output-modifying-arguments)
   - [Content-modifying arguments](#content-modifying-arguments)
   - [Using include/exclude and redirects](#using-include/exclude-and-redirects)
 - [Response objects](#response-objects)
   - [Post response object](#post-response-object)
   - [Category response object](#category-response-object)
   - [Tag response object](#tag-response-object)
   - [Author response object](#author-response-object)
   - [Comment response object](#comment-response-object)
   - [Attachment response object](#attachment-response-object)
 - [Extending JSON API](#extending-json-api)
   - [Plugin hooks](#plugin-hooks)
   - [Developing JSON API controllers](#developing-json-api-controllers)
   - [Configuration options](#configuration-options)
 - [Unit tests](#unit-tests)
   - [Preparing a WordPress test site](#preparing-a-wordpress-test-site)
   - [Running the tests](#running-the-tests)

##General Concepts
###Requests
Requests use a simple HTTP GET or POST. To invoke the API, include a non-empty query value for `json` in the URL.

JSON API operates in two modes:

1. *Implicit mode* is triggered by setting the `json` query var to a non-empty value on any WordPress page. The content that would normally appear on that page is returned in JSON format.
2. *Explicit mode* is triggered by setting `json` to a known method string. See [Request methods](#request-methods) for a complete method listing.

#####Implicit mode examples:

 - `http://www.example.org/?json=1`
 - `http://www.example.org/?p=47&json=1`
 - `http://www.example.org/tag/banana/?json=1`

#####Explicit mode examples:

 - `http://www.example.org/?json=get_recent_posts`
 - `http://www.example.org/?json=get_post&post_id=47`
 - `http://www.example.org/?json=get_tag_posts&tag_slug=banana`

#####With user-friendly permalinks configured:

 - `http://www.example.org/api/get_recent_posts/`
 - `http://www.example.org/api/get_post/?post_id=47`
 - `http://www.example.org/api/get_tag_posts/?tag_slug=banana`

__Further reading__
See [Request arguments](#request-arguments) for more information about request arguments to modify the response.

###Controllers
The 1.0 release of JSON API introduced a modular controller system. This allows developers to flexibly add features to the API and give users more control over which methods they have enabled.

#####The Core controller
Most of the methods available prior to version 1.0 have been moved to the Core controller. The two exceptions are `submit_comment` and `create_post` which are now available from the Respond and Posts controllers, respectively. The Core controller is the only one enabled by default. All other functionality must be enabled from the JSON API Settings page (under Settings in the WordPress admin menu).

#####Specifying a controller
There are a few ways of specifying a controller, depending on how you are calling the API:

 - `http://www.example.org/?json=get_recent_posts` (`core` controller is implied, method is `get_recent_posts`)
 - `http://www.example.org/api/info/` (`core` controller is implied)
 - `http://www.example.org/api/core/get_category_posts/` (`core` controller can also be explicitly specified)
 - `http://www.example.org/?json=respond.submit_comment` (`respond` controller, `submit_comment` method)

__Legacy compatibility__
JSON API retains support for its pre-1.0 methods. For example, if you invoke the method `create_post` without a controller specified, the Posts controller is chosen instead of Core.

#####Available controllers
The current release includes three controllers:
 - Core
 - Posts
 - Respond

Developers are encouraged to suggest or submit additional controllers.

__Further reading__
See [Request methods](#request-methods) for a complete reference of available controllers and methods. For documentation on extending JSON API with new controllers see [Developing JSON API controllers](#developing-json-api-controllers).

###Responses
The standard response format for JSON API is (as you may have guessed) [JSON](http://json.org/).

Here is an example response from `http://localhost/wordpress/?json=1` called on a default WordPress installation (formatted for readability):

```json
{
  "status": "ok",
  "count": 1,
  "count_total": 1,
  "pages": 1,
  "posts": [
    {
      "id": 1,
      "type": "post",
      "slug": "hello-world",
      "url": "http:\/\/localhost\/wordpress\/?p=1",
      "title": "Hello world!",
      "title_plain": "Hello world!",
      "content": "<p>Welcome to WordPress. This is your first post. Edit or delete it, then start blogging!<\/p>\n",
      "excerpt": "Welcome to WordPress. This is your first post. Edit or delete it, then start blogging!\n",
      "date": "2009-11-11 12:50:19",
      "modified": "2009-11-11 12:50:19",
      "categories": [],
      "tags": [],
      "author": {
        "id": 1,
        "slug": "admin",
        "name": "admin",
        "first_name": "",
        "last_name": "",
        "nickname": "",
        "url": "",
        "description": ""
      },
      "comments": [
        {
          "id": 1,
          "name": "Mr WordPress",
          "url": "http:\/\/wordpress.org\/",
          "date": "2009-11-11 12:50:19",
          "content": "<p>Hi, this is a comment.<br \/>To delete a comment, just log in and view the post&#039;s comments. There you will have the option to edit or delete them.<\/p>\n",
          "parent": 0
        }
      ],
      "comment_count": 1,
      "comment_status": "open"
    }
  ]
}
```

##Request methods
Request methods are available from the following controllers:

 - Core controller - basic introspection methods
 - Posts controller - data manipulation methods for posts
 - Respond controller - comment/trackback submission methods
 - Widgets controller - retrieve sidebar widgets

###Core controller methods
The Core controller offers a mostly-complete set of introspection methods for retrieving content from WordPress.

####Method: `info`
Returns information about JSON API.

#####Optional arguments
 - `controller` - returns detailed information about a specific controller

#####Response
```json
{
  "status": "ok",
  "json_api_version": "1.0",
  "controllers": [
    "core"
  ]
}
```

#####Response with “controller=core”
```json
{
  "status": "ok",
  "name": "Core",
  "description": "Basic introspection methods",
  "methods": [
    ...
  ]
}
```

####Method: `get_recent_posts`
Returns an array of recent posts. You can invoke this from the WordPress home page either by setting `json` to a non-empty value (i.e., `json=1`) or from any page by setting `json=get_recent_posts`.

#####Optional arguments
 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "count": 10,
  "count_total": 79,
  "pages": 7,
  "posts": [
    { ... },
    { ... },
    ...
  ]
}
```

####Method: `get_posts`
Returns posts according to WordPress's [`WP_Query` parameters](http://codex.wordpress.org/Class_Reference/WP_Query#Parameters). The one default parameter is `ignore_sticky_posts=1` (this can be overridden).

#####Optional arguments

 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

__Further reading__
See the [`WP_Query` documentation](http://codex.wordpress.org/Class_Reference/WP_Query#Parameters) for a full list of supported parameters. The `post_status` parameter is currently ignored.

#####Response
```json
{
  "status": "ok",
  "count": 1,
  "posts": [
    { ... }
  ]
}
```

####Method: `get_post`
Returns a single post object.

#####One of the following is required
 - Invoking the JSON API implicitly (i.e., `?json=1`) on a post URL
 - `id` or `post_id` - set to the post's ID
 - `slug` or `post_slug` - set to the post's URL slug

#####Optional arguments
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "post": { ... }
}
```

####Method: `get_page`
Returns a single page object.

#####One of the following is required
 - Invoking the JSON API implicitly (i.e., `?json=1`) on a page URL
 - `id` or `page_id` - set to the page's ID
 - `slug` or `page_slug` - set to the page's URL slug

#####Optional arguments
 - `children` - set to a non-empty value to include a recursive hierarchy of child pages
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "page": { ... }
}
```

####Method: `get_date_posts`
Returns an array of posts/pages in a specific date archive (by day, month, or year).

#####One of the following is required
 - Invoking the JSON API implicitly (i.e., `?json=1`) on a date archive page
 - `date` - set to a date in the format `YYYY` or `YYYY-MM` or `YYYY-MM-DD` (non-numeric characters are stripped from the var, so `YYYYMMDD` or `YYYY/MM/DD` are also valid)

#####Optional arguments

 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "count": 10,
  "count_total": 79,
  "pages": 7,
  "posts": [
    { ... },
    { ... },
    ...
  ]
}
```

####Method: `get_category_posts`
Returns an array of posts/pages in a specific category.

#####One of the following is required

 - Invoking the JSON API implicitly (i.e., `?json=1`) on a category archive page
 - `id` or `category_id` - set to the category's ID
 - `slug` or `category_slug` - set to the category's URL slug

#####Optional arguments
 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "count": 10,
  "count_total": 79,
  "pages": 7,
  "category": { ... }
  "posts": [
    { ... },
    { ... },
    ...
  ]
}
```

####Method: `get_tag_posts`
Returns an array of posts/pages with a specific tag.

#####One of the following is required
 - Invoking the JSON API implicitly (i.e., `?json=1`) on a tag archive page
 - `id` or `tag_id` - set to the tag's ID
 - `slug` or `tag_slug` - set to the tag's URL slug

#####Optional arguments
 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "count": 10,
  "count_total": 79,
  "pages": 7,
  "tag": { ... }
  "posts": [
    { ... },
    { ... },
    ...
  ]
}
```

####Method: `get_author_posts`
Returns an array of posts/pages written by a specific author.

#####One of the following is required
 - Invoking the JSON API implicitly (i.e., `?json=1`) on an author archive page
 - `id` or `author_id` - set to the author's ID
 - `slug` or `author_slug` - set to the author's URL slug

#####Optional arguments
 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "count": 10,
  "count_total": 79,
  "pages": 7,
  "author": { ... }
  "posts": [
    { ... },
    { ... },
    ...
  ]
}
```

####Method: `get_search_results`
Returns an array of posts/pages in response to a search query.

#####One of the following is required

 - Invoking the JSON API implicitly (i.e., `?json=1`) on a search results page
 - `search` - set to the desired search query

#####Optional arguments
 - `count` - determines how many posts per page are returned (default value is 10)
 - `page` - return a specific page number from the results
 - `post_type` - used to retrieve custom post types

#####Response
```json
{
  "status": "ok",
  "count": 10,
  "count_total": 79,
  "pages": 7,
  "posts": [
    { ... },
    { ... },
    ...
  ]
}
```

####Method: `get_date_index`
Returns both an array of date page permalinks and a tree structure representation of the archive.

#####Response
```json
{
  "status": "ok",
  "permalinks": [
    "...",
    "...",
    "..."
  ],
  "tree": {
    "2009": {
      "09": 17,
      "10": 20,
      "11": 7
    }
  }
}
```

Note: the tree is arranged by `response.tree.[year].[month].[number of posts]`.

####Method: `get_category_index`
Returns an array of active categories.

#####Optional argument
 - `parent` - returns categories that are direct children of the parent ID

#####Response
```json
{
  "status": "ok",
  "count": 3,
  "categories": [
    { ... },
    { ... },
    { ... }
  ]
}
```

####Method: `get_tag_index`
Returns an array of active tags.

#####Response
```json
{
  "status": "ok",
  "count": 3,
  "tags": [
    { ... },
    { ... },
    { ... }
  ]
}
```

####Method: `get_author_index`
Returns an array of active blog authors.

#####Response
```json
{
  "status": "ok",
  "count": 3,
  "authors": [
    { ... },
    { ... },
    { ... }
  ]
}
```

####Method: `get_page_index`
Returns a hierarchical tree of `page` posts.

#####Response
```json
{
  "status": "ok",
  "pages": [
    { ... },
    { ... },
    { ... }
  ]
}
```

####Method: `get_nonce`
Returns a WordPress nonce value, required to call some data manipulation methods.

#####Required arguments
 - `controller` - the JSON API controller for the method you will use the nonce for
 - `method` - the method you wish to call (currently `create_post` is the only method that requires a nonce)

#####Response
```json
{
  "status": "ok",
  "controller": "posts",
  "method": "create_post",
  "nonce": "cefe01efd4"
}
```

__Further reading__
To learn more about how nonces are used in WordPress, see [Mark Jaquith's article on the subject](http://markjaquith.wordpress.com/2006/06/02/wordpress-203-nonces/).

###Pages controller methods
####Method: `create_post`
Creates a new post.

#####Required argument
 - `nonce` - available from the `get_nonce` method (call with vars `controller=posts` and `method=create_post`)

#####Optional arguments
 - `status` - sets the post status ("draft" or "publish"), default is "draft"
 - `title` - the post title
 - `content` - the post content
 - `author` - the post's author (login name), default is the current logged in user
 - `categories` - a comma-separated list of categories (URL slugs)
 - `tags` - a comma-separated list of tags (URL slugs)

Note: including a file upload field called `attachment` will cause an attachment to be stored with your new post.

####Method: `update_post`

History.md
#Changelog
##1.1.1 (2013-06-23):
 - Added support for custom taxonomies
 - Errors are now suppressed unless you include a non-empty `dev` argument

##1.1.0 (2013-06-22):
 - Bugfix for `json_encode` compatibility with PHP < 5.3
 - Bugfix for `get_author_index` warnings in WordPress > 3.5

##1.0.9 (2013-06-21):
 - Added `update_post` and `delete_post` methods to Post controller
 - Added two JSON encoding arguments: `json_encode_options` and `json_unescaped_unicode`
 - Added a `parent` argument to `get_category_index`
 - Fixed a couple places where the code was generating PHP notifications
 - Updated bundled Services_JSON library (only used if `json_encode` is unavailable)

##1.0.8 (2013-06-12):
 - Added `widgets` controller
 - Added a generic `get_posts` method to the core controller
 - Added a `thumbnail_images` object property to complement `thumbnail` URL
 - Attachment image files are now checked to exist and match the expected width/height
 - Fixed a bug where `the_excerpt` filter wasn't being applied to the `excerpt` property
 - Fixed a bug where the number of child pages was being limited to 5
 - Fixed a bug where custom controller class names couldn't include numerics
 - Theme directory check for custom controllers

##1.0.7 (2011-01-27):
 - Created some basic unit tests
 - Fixed a bug where `get_author_posts` was unable to find users by `slug`
 - Added missing `post_type` argument to documentation for `get_post` and `get_page` (props Koshirosan)
 - Added `previous_url` and `next_url` properties to the `get_post` response object (props mlcy44)

##1.0.6 (2011-01-13):
 - Fixed a bug in `exclude` query parameter (big props to ikesyo and archon810)
 - Fix for `get_page_index` that where it only returned 5 pages -- it now responds to `count` query param (props to npavkovic and blinder)
 - Removed `Content-Disposition` header from response (props mimecine, kjwierenga)
 - Fixed an incompatibility issue with Disqus plugin (props joshcanhelp)
 - Fixed a bug where `submit_comment` was resulting in a HTTP 404 status (props @tdweston)
 - Fixed an error in the documentation, external controller example (props jli)

##1.0.5 (2010-07-08):
 - Added an check so that `json-api.php` can be moved one level above the `json-api` directory
 - Added more documentation about using nonces

##1.0.4 (2010-07-07):
 - Fixed a bug where the order of attachments didn't match the gallery
 - Added a section to the developer documentation for externalizing custom controllers
 - Moved JSON_API class to its own file: `singletons/api.php`
 - Created a new top-level function: `json_api_dir()`
 - Improvements for WordPress MU: `JSON_API_DIR` and `JSON_API_CONTROLLERS` constants (props Jim McQuillan)

##1.0.3 (2010-07-07):
 - Added request argument `thumbnail_size` to support different sizes of featured images (see also: `add_image_size` WordPress function)
 - Added request argument `post_type` to support custom post types (props Mark Harris)

##1.0.2 (2010-07-02):
 - Removed an inaccurate section from readme.txt about supporting `query_posts` arguments
 - Changed controller info block format to use "Controller name" and "Controller description"
 - Made admin page more robust about handling errors loading controllers
 - Changed `JSON_API::get_controllers` method to lowercase all entries
 - Added introspector section to developer documentation
 - Fixed incorrect example for `json_api_[controller]_controller_path`
 - Thanks to Tim Nash for early feedback on writing external controllers

##1.0.1 (2010-07-01):
 - Fixed some typos in readme.txt
 - Switched `get_tag_posts` to query on tag instead of tag_id (maybe a WordPress issue?)

##1.0 (2010-06-29):
 - JSON API officially drops support for PHP 4 (it was already broken)
 - Added JSON API Settings page to WP admin
 - Broke apart `JSON_API_Controller` into a modular controller system
 - Refactored `JSON_API_Query` to depend less on WordPress's `get_query_var` mechanism
 - Developer mode now shows response in JSON format
 - The `create_post` method now requires a nonce
 - Improved support for complex post queries (props zibitt)
 - Fixed a bug with `get_author_by_login` (props Krzysztof Sobolewski)
 - Made image attachments more robust with `get_intermediate_image_sizes` (props mimecine)
 - Improved post thumbnail support (props nyamsprod)

##0.9.6 (2010-05-27):
 - Fixed a bug introduced in 0.9.5

##0.9.5 (2010-05-27):
 - Added a `thumbnail` property to Post response objects

##0.9.4 (2010-04-28):
 - Fixed a bug where any non-authenticated user could create a draft blog post through `create_post`. Thanks to user futtta for posting about this.

##0.9.3 (2010-03-19):
 - Fixed a bug where child pages were being ignored by the API. See also: https://core.trac.wordpress.org/ticket/12647

##0.9.2 (2010-03-18):
 - Fixed a bug where the /api/ rewrite rules would be lost

##0.9 (2010-02-04):
 - Added a `create_post` method

##0.8.3 (2010-01-27):
 - Fixed the stable tag version

##0.8.2 (2010-01-27):
 - Fixed a typo in the changelog

##0.8.1 (2010-01-27):
 - Fixed a bug that was making JSONP support non-functional

##0.8 (2010-01-18):
 - Added an attachment model and instance variable for post objects

##0.7.3 (2010-01-15):
 - Added a `count` request parameter to control the number of posts returned

##0.7.2 (2010-01-14):
 - Removed the version number from the description text

##0.7.1 (2010-01-14):
 - Fixed another subtle bug with `get_author_index`

##0.7 (2010-01-08):
 - Added a `post_count` response to tag objects
 - Fixed a bug with `get_author_index`

##0.6 (2009-11-30):
 - Added `count_total` response
 - Added `json_api_encode` filter
 - Fixed bugs in the introspector's `get_current_category` and `get_current_tag`

##0.5 (2009-11-17):
 - Initial Public Release

#Upgrade Notice

##1.1.1
Added support for custom taxonomies

##1.1.0
Minor bugfixes

##1.0.9
Update/delete post methods and some other bugfixes and improvements

##1.0.8
Long overdue bugfix/improvement release

##1.0.7

singletons/api.php
<?php

class JSON_API {
  
  function __construct() {
    $this->query = new JSON_API_Query();
    $this->introspector = new JSON_API_Introspector();
    $this->response = new JSON_API_Response();
    add_action('template_redirect', array(&$this, 'template_redirect'));
    add_action('admin_menu', array(&$this, 'admin_menu'));
    add_action('update_option_json_api_base', array(&$this, 'flush_rewrite_rules'));
    add_action('pre_update_option_json_api_controllers', array(&$this, 'update_controllers'));
  }
  
  function template_redirect() {
    // Check to see if there's an appropriate API controller + method    
    $controller = strtolower($this->query->get_controller());
    $available_controllers = $this->get_controllers();
    $enabled_controllers = explode(',', get_option('json_api_controllers', 'core'));
    $active_controllers = array_intersect($available_controllers, $enabled_controllers);
    
    if ($controller) {
      
      if (empty($this->query->dev)) {
        error_reporting(0);
      }
      
      if (!in_array($controller, $active_controllers)) {
        $this->error("Unknown controller '$controller'.");
      }
      
      $controller_path = $this->controller_path($controller);
      if (file_exists($controller_path)) {
        require_once $controller_path;
      }
      $controller_class = $this->controller_class($controller);
      
      if (!class_exists($controller_class)) {
        $this->error("Unknown controller '$controller_class'.");
      }
      
      $this->controller = new $controller_class();
      $method = $this->query->get_method($controller);
      
      if ($method) {
        
        $this->response->setup();
        
        // Run action hooks for method
        do_action("json_api", $controller, $method);
        do_action("json_api-{$controller}-$method");
        
        // Error out if nothing is found
        if ($method == '404') {
          $this->error('Not found');
        }
        
        // Run the method
        $result = $this->controller->$method();
        
        // Handle the result
        $this->response->respond($result);
        
        // Done!
        exit;
      }
    }
  }
  
  function admin_menu() {
    add_options_page('JSON API Settings', 'JSON API', 'manage_options', 'json-api', array(&$this, 'admin_options'));
  }
  
  function admin_options() {
    if (!current_user_can('manage_options'))  {
      wp_die( __('You do not have sufficient permissions to access this page.') );
    }
    
    $available_controllers = $this->get_controllers();
    $active_controllers = explode(',', get_option('json_api_controllers', 'core'));
    
    if (count($active_controllers) == 1 && empty($active_controllers[0])) {
      $active_controllers = array();
    }
    
    if (!empty($_REQUEST['_wpnonce']) && wp_verify_nonce($_REQUEST['_wpnonce'], "update-options")) {
      if ((!empty($_REQUEST['action']) || !empty($_REQUEST['action2'])) &&
          (!empty($_REQUEST['controller']) || !empty($_REQUEST['controllers']))) {
        if (!empty($_REQUEST['action'])) {
          $action = $_REQUEST['action'];
        } else {
          $action = $_REQUEST['action2'];
        }
        
        if (!empty($_REQUEST['controllers'])) {
          $controllers = $_REQUEST['controllers'];
        } else {
          $controllers = array($_REQUEST['controller']);
        }
        
        foreach ($controllers as $controller) {
          if (in_array($controller, $available_controllers)) {
            if ($action == 'activate' && !in_array($controller, $active_controllers)) {
              $active_controllers[] = $controller;
            } else if ($action == 'deactivate') {
              $index = array_search($controller, $active_controllers);
              if ($index !== false) {
                unset($active_controllers[$index]);
              }
            }
          }
        }
        $this->save_option('json_api_controllers', implode(',', $active_controllers));
      }
      if (isset($_REQUEST['json_api_base'])) {
        $this->save_option('json_api_base', $_REQUEST['json_api_base']);
      }
    }
    
    ?>
<div class="wrap">
  <div id="icon-options-general" class="icon32"><br /></div>
  <h2>JSON API Settings</h2>
  <form action="options-general.php?page=json-api" method="post">
    <?php wp_nonce_field('update-options'); ?>
    <h3>Controllers</h3>
    <?php $this->print_controller_actions(); ?>
    <table id="all-plugins-table" class="widefat">
      <thead>
        <tr>
          <th class="manage-column check-column" scope="col"><input type="checkbox" /></th>
          <th class="manage-column" scope="col">Controller</th>
          <th class="manage-column" scope="col">Description</th>
        </tr>
      </thead>
      <tfoot>
        <tr>
          <th class="manage-column check-column" scope="col"><input type="checkbox" /></th>
          <th class="manage-column" scope="col">Controller</th>
          <th class="manage-column" scope="col">Description</th>
        </tr>
      </tfoot>
      <tbody class="plugins">
        <?php
        
        foreach ($available_controllers as $controller) {
          
          $error = false;
          $active = in_array($controller, $active_controllers);
          $info = $this->controller_info($controller);
          
          if (is_string($info)) {
            $active = false;
            $error = true;
            $info = array(
              'name' => $controller,
              'description' => "<p><strong>Error</strong>: $info</p>",
              'methods' => array(),
              'url' => null
            );
          }
          
          ?>
          <tr class="<?php echo ($active ? 'active' : 'inactive'); ?>">
            <th class="check-column" scope="row">
              <input type="checkbox" name="controllers[]" value="<?php echo $controller; ?>" />
            </th>
            <td class="plugin-title">
              <strong><?php echo $info['name']; ?></strong>
              <div class="row-actions-visible">
                <?php
                
                if ($active) {
                  echo '<a href="' . wp_nonce_url('options-general.php?page=json-api&amp;action=deactivate&amp;controller=' . $controller, 'update-options') . '" title="' . __('Deactivate this controller') . '" class="edit">' . __('Deactivate') . '</a>';
                } else if (!$error) {
                  echo '<a href="' . wp_nonce_url('options-general.php?page=json-api&amp;action=activate&amp;controller=' . $controller, 'update-options') . '" title="' . __('Activate this controller') . '" class="edit">' . __('Activate') . '</a>';
                }
                  
                if (!empty($info['url'])) {
                  echo ' | ';
                  echo '<a href="' . $info['url'] . '" target="_blank">Docs</a></div>';
                }
                
                ?>
            </td>
            <td class="desc">
              <p><?php echo $info['description']; ?></p>
              <p>
                <?php
                
                foreach($info['methods'] as $method) {
                  $url = $this->get_method_url($controller, $method);
                  if ($active) {
                    echo "<code><a href=\"$url\">$method</a></code> ";
                  } else {
                    echo "<code>$method</code> ";
                  }
                }
                
                ?>
              </p>
            </td>
          </tr>
        <?php } ?>
      </tbody>
    </table>
    <?php $this->print_controller_actions('action2'); ?>
    <h3>Address</h3>
    <p>Specify a base URL for JSON API. For example, using <code>api</code> as your API base URL would enable the following <code><?php bloginfo('url'); ?>/api/get_recent_posts/</code>. If you assign a blank value the API will only be available by setting a <code>json</code> query variable.</p>
    <table class="form-table">
      <tr valign="top">
        <th scope="row">API base</th>
        <td><code><?php bloginfo('url'); ?>/</code><input type="text" name="json_api_base" value="<?php echo get_option('json_api_base', 'api'); ?>" size="15" /></td>
      </tr>
    </table>
    <?php if (!get_option('permalink_structure', '')) { ?>
      <br />
      <p><strong>Note:</strong> User-friendly permalinks are not currently enabled. <a target="_blank" class="button" href="options-permalink.php">Change Permalinks</a>
    <?php } ?>
    <p class="submit">
      <input type="submit" class="button-primary" value="<?php _e('Save Changes') ?>" />
    </p>
  </form>
</div>
<?php
  }
  
  function print_controller_actions($name = 'action') {
    ?>
    <div class="tablenav">
      <div class="alignleft actions">
        <select name="<?php echo $name; ?>">
          <option selected="selected" value="-1">Bulk Actions</option>
          <option value="activate">Activate</option>
          <option value="deactivate">Deactivate</option>
        </select>
        <input type="submit" class="button-secondary action" id="doaction" name="doaction" value="Apply">
      </div>
      <div class="clear"></div>
    </div>
    <div class="clear"></div>
    <?php
  }
  
  function get_method_url($controller, $method, $options = '') {
    $url = get_bloginfo('url');
    $base = get_option('json_api_base', 'api');
    $permalink_structure = get_option('permalink_structure', '');
    if (!empty($options) && is_array($options)) {
      $args = array();
      foreach ($options as $key => $value) {
        $args[] = urlencode($key) . '=' . urlencode($value);
      }
      $args = implode('&', $args);
    } else {
      $args = $options;
    }
    if ($controller != 'core') {
      $method = "$controller/$method";
    }
    if (!empty($base) && !empty($permalink_structure)) {
      if (!empty($args)) {
        $args = "?$args";
      }
      return "$url/$base/$method/$args";
    } else {
      return "$url?json=$method&$args";
    }
  }
  
  function save_option($id, $value) {
    $option_exists = (get_option($id, null) !== null);
    if ($option_exists) {
      update_option($id, $value);
    } else {
      add_option($id, $value);
    }
  }
  
  function get_controllers() {
    $controllers = array();
    $dir = json_api_dir();
    $this->check_directory_for_controllers("$dir/controllers", $controllers);
    $this->check_directory_for_controllers(get_stylesheet_directory(), $controllers);
    $controllers = apply_filters('json_api_controllers', $controllers);
    return array_map('strtolower', $controllers);
  }
  
  function check_directory_for_controllers($dir, &$controllers) {
    $dh = opendir($dir);
    while ($file = readdir($dh)) {
      if (preg_match('/(.+)\.php$/i', $file, $matches)) {
        $src = file_get_contents("$dir/$file");
        if (preg_match("/class\s+JSON_API_{$matches[1]}_Controller/i", $src)) {
          $controllers[] = $matches[1];
        }
      }
    }
  }
  
  function controller_is_active($controller) {
    if (defined('JSON_API_CONTROLLERS')) {
      $default = JSON_API_CONTROLLERS;
    } else {
      $default = 'core';
    }
    $active_controllers = explode(',', get_option('json_api_controllers', $default));
    return (in_array($controller, $active_controllers));
  }
  
  function update_controllers($controllers) {
    if (is_array($controllers)) {
      return implode(',', $controllers);
    } else {
      return $controllers;
    }
  }
  
  function controller_info($controller) {
    $path = $this->controller_path($controller);
    $class = $this->controller_class($controller);
    $response = array(
      'name' => $controller,
      'description' => '(No description available)',
      'methods' => array()
    );
    if (file_exists($path)) {
      $source = file_get_contents($path);
      if (preg_match('/^\s*Controller name:(.+)$/im', $source, $matches)) {
        $response['name'] = trim($matches[1]);
      }
      if (preg_match('/^\s*Controller description:(.+)$/im', $source, $matches)) {
        $response['description'] = trim($matches[1]);
      }
      if (preg_match('/^\s*Controller URI:(.+)$/im', $source, $matches)) {
        $response['docs'] = trim($matches[1]);
      }
      if (!class_exists($class)) {
        require_once($path);
      }
      $response['methods'] = get_class_methods($class);
      return $response;
    } else if (is_admin()) {
      return "Cannot find controller class '$class' (filtered path: $path).";
    } else {
      $this->error("Unknown controller '$controller'.");
    }
    return $response;
  }
  
  function controller_class($controller) {
    return "json_api_{$controller}_controller";
  }
  
  function controller_path($controller) {
    $json_api_dir = json_api_dir();
    $json_api_path = "$json_api_dir/controllers/$controller.php";
    $theme_dir = get_stylesheet_directory();
    $theme_path = "$theme_dir/$controller.php";
    if (file_exists($theme_path)) {
      $path = $theme_path;
    } else if (file_exists($json_api_path)) {
      $path = $json_api_path;
    } else {
      $path = null;
    }
    $controller_class = $this->controller_class($controller);
    return apply_filters("{$controller_class}_path", $path);
  }
  
  function get_nonce_id($controller, $method) {
    $controller = strtolower($controller);
    $method = strtolower($method);
    return "json_api-$controller-$method";
  }
  
  function flush_rewrite_rules() {
    global $wp_rewrite;
    $wp_rewrite->flush_rules();
  }
  
  function error($message = 'Unknown error', $status = 'error', $header = '400') {
    $this->response->respond(array(
      'error' => $message
    ), $status, $header);
  }
  
  function include_value($key) {
    return $this->response->is_value_included($key);
  }
  
}

?>
